test/functional/feature_smart_contract_test.py

```diff
@@ -0,0 +1,144 @@
+from collections import defaultdict
+
+from substrateinterface import Keypair
+import test_framework.mintlayer.utxo as utxo
+import test_framework.mintlayer.contract as contract
+
+from test_framework.test_framework import MintlayerTestFramework
+from test_framework.util import (
+    assert_equal,
+    connect_nodes,
+    wait_until,
+)
+
+import os
+
+
+class ExampleTest(MintlayerTestFramework):
+    # Each functional test is a subclass of the MintlayerTestFramework class.
+
+    # Override the set_test_params(), add_options(), setup_chain(), setup_network()
+    # and setup_nodes() methods to customize the test setup as required.
+
+    def set_test_params(self):
+        """Override test parameters for your individual test.
+
+        This method must be overridden and num_nodes must be exlicitly set."""
+        self.setup_clean_chain = True
+        self.num_nodes = 3
+        # Use self.extra_args to change command-line arguments for the nodes
+        self.extra_args = [[], [], []]
+
+        # self.log.info("I've finished set_test_params")  # Oops! Can't run self.log before run_test()
+
+    def setup_network(self):
+        """Setup the test network topology
+
+        Often you won't need to override this, since the standard network topology
+        (linear: node0 <-> node1 <-> node2 <-> ...) is fine for most tests.
+
+        If you do override this method, remember to start the nodes, assign
+        them to self.nodes, connect them and then sync."""
+
+        self.setup_nodes()
+
+        # In this test, we're not connecting node2 to node0 or node1. Calls to
+        # sync_all() should not include node2, since we're not expecting it to
+        # sync.
+        connect_nodes(self.nodes[0], self.nodes[1])
+        # self.sync_all([self.nodes[0:1]])
+
+    def run_test(self):
+        """Main test logic"""
+        client = self.nodes[0].rpc_client
+
+        substrate = client.substrate
+
+        alice = Keypair.create_from_uri('//Alice')
+
+        # Find a suitable UTXO
+        initial_utxo = [x for x in client.utxos_for(alice) if x[1].value >= 50][0]
+
+        tx0 = utxo.Transaction(
+            client,
+            inputs=[
+                utxo.Input(initial_utxo[0]),
+            ],
+            outputs=[
+                utxo.Output(
+                    value=50,
+                    header=0,
+                    destination=utxo.DestPubkey(alice.public_key)
+                ),
+                utxo.Output(
+                    value=0,
+                    header=0,
+                    destination=utxo.DestCreatePP(
+                        code=os.path.join(os.path.dirname(__file__), "code.wasm"),
+                        data=[0xed, 0x4b, 0x9d, 0x1b],  # default() constructor selector
+                    )
+                ),
+            ]
+        ).sign(alice, [initial_utxo[1]])
+
+        # submit transaction and get the extrinsic and block hashes
+        (ext, blk) = client.submit(alice, tx0)
+
```

Comment:
One thing we could also do at some point is to verify that the UTXO set contains correct UTXOs. Like here we could test that in addition to the UTXOs it had before anything was executed, we could check that it also contains UTXOs with values 0 and 50.